### PR TITLE
Fibonacci: memoization approach

### PR DESCRIPTION
**Pull request by Femchengdu**
The Fibonacci numbers, commonly denoted `F(n)` form a sequence, called the Fibonacci sequence, such that each number is the sum of the two preceding ones, starting from `0` and `1`. That is,

```
F(0) = 0, F(1) = 1
F(n) = F(n - 1) + F(n - 2), for n > 1.
```

Given n, calculate `F(n)`.

### CODE REVIEW

**Review by vzvu3k6k — COMMENTED** [review]
Looks good. (I like the way of use of a default argument `memo_hash` 👍 )

But I think the filename should be `dynamic_programming/fibonacci.rb`, not `fast_fibonacci.rb`. This DP algorithm is O(n) and is certainly faster than a simple recursion, but there is a much faster algorithm of O(log n). Other repositories call it as fast.

- https://github.com/TheAlgorithms/C/blob/master/misc/fibonacci_fast.c
- https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/fibonacci_fast.cpp
- https://github.com/TheAlgorithms/Python/blob/master/dynamic_programming/fast_fibonacci.py

**Comment by vbrazo** [comment]
@vzvu3k6k I think you're talking about a different approach. The way how I see it is that he added the `Bottom-Up Approach using Memoization` and yes it's not the fastest one so I'd also suggest renaming your method. The fastest approach is the `Iterative Top-Down Approach` which has time complexity `O(n)` and space complexity `O(1)`. This could be included in a separate PR.

File: dynamic_programming/fast_fibonacci.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Given a number, calculates the nth fibonacci number.
+
+def fast_fibonacci(number, memo_hash = {})
```
Comment (vbrazo) [inline dynamic_programming/fast_fibonacci.rb:3]:
```suggestion
# Given a number, calculates the nth Fibonacci number.

#
# Approach: Bottom-Up Approach using Memoization
#

def fibonacci(number, memo_hash = {})
```

**Comment by vbrazo** [comment]
Actually the fastest one is using Math:

```ruby
#
# Approach 5: Math
#

# Intuition Using the golden ratio, a.k.a Binet's forumula

# Algorithm
#
# Use the golden ratio formula to calculate the Nth Fibonacci number.
# https://demonstrations.wolfram.com/GeneralizedFibonacciSequenceAndTheGoldenRatio/

# Complexity Analysis

# Time complexity: O(1). Constant time complexity since we are using no loops or recursion and the time is based on the result of performing the calculation using Binet's formula.
# Space complexity: O(1). The space used is the space needed to create the variable to store the golden ratio formula.

def fib(n)
  golden_ratio = (1 + 5**0.5) / 2
  ((golden_ratio**n + 1) / 5**0.5).to_i
end

n = 2
# Output: 1
# Explanation: F(2) = F(1) + F(0) = 1 + 0 = 1.
puts(fib(n))
```
I'll include in a new PR.

**Comment by vbrazo** [comment]
@Femchengdu @vzvu3k6k please review: https://github.com/TheAlgorithms/Ruby/pull/92

**Review by vbrazo — APPROVED** [review]
@Femchengdu I added a few changes to your PR if you don't mind. Thanks for the contributions 👍

**Comment by Femchengdu** [comment]
@vbrazo I appreciate your feedback, thank you.

**Pull request merged**